### PR TITLE
Allow removal of enum declarations from contract interfaces for C1.0 upgrade

### PR DESCRIPTION
**Pull request by dsainati1**
Closes https://github.com/onflow/cadence/issues/3182

This allows enum declarations to be removed from contract interfaces as part of the Cadence 1.0 upgrade. While in general this is not a safe operation to permit, analysis of mainnet contracts indicates that there are a small number of existing declarations of this kind, none of which are used in fields that themselves cannot be easily removed. 

______

- [X] Targeted PR against `master` branch
- [X] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [X] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [X] Updated relevant documentation 
- [X] Re-reviewed `Files changed` in the Github PR explorer
- [X] Added appropriate labels

### CODE REVIEW

File: runtime/stdlib/cadence_v0.42_to_v1_contract_upgrade_validator.go [hunk 1]
```diff
@@ -533,6 +533,27 @@ func (validator *CadenceV042ToV1ContractUpdateValidator) checkDeclarationKindCha
 	return false
 }
 
+func (validator *CadenceV042ToV1ContractUpdateValidator) checkNestedDeclarationRemoval(
```
Comment (turbolent) [inline runtime/stdlib/cadence_v0.42_to_v1_contract_upgrade_validator.go:536]: Nice!

**Review by turbolent — APPROVED** [review]
Looks good!

Could you please also add or check we already have a test that it is not possible to add a new enum declaration?

File: runtime/stdlib/cadence_v0.42_to_v1_contract_upgrade_validation_test.go [hunk 2]
```diff
@@ -2066,3 +2073,51 @@ func TestInterfaceConformanceChange(t *testing.T) {
...
+            }
+        `
+
+		const newCode = `
+            access(all) contract Test {}
+        `
+
+		err := testContractUpdate(t, oldCode, newCode)
+
+		// This is not allowed because `@R{I}` is converted to `@R`, not `@{I}`
```
Comment (SupunS) [inline runtime/stdlib/cadence_v0.42_to_v1_contract_upgrade_validation_test.go:2097]:
```suggestion
```

**Review by SupunS — APPROVED** [review]

**Comment by dsainati1** [comment]
> Could you please also add or check we already have a test that it is not possible to add a new enum declaration?

Why should this not be possible? This should be allowed even by the regular validator.

**Comment by turbolent** [comment]
> > Could you please also add or check we already have a test that it is not possible to add a new enum declaration?
> 
> Why should this not be possible? This should be allowed even by the regular validator.

We can only allow the removal if we also forbid the addition, otherwise an incompatible declaration could be reintroduced, circumventing the update checking rules for enums.

**Comment by dsainati1** [comment]
Oh, yes in that case the new type checking will prevent enums from being added to interfaces implicitly; those contracts won't type check so there's no need to forbid it in the validator.

**Pull request merged**